stats/opentelemetry/internal/tracing/custom_carrier.go

```diff
@@ -61,22 +73,27 @@ func (c *CustomCarrier) Set(key, value string) {
 }
 
 // Keys returns the keys stored in the carrier's context metadata. It returns
-// keys from both outgoing and incoming context metadata.
+// keys from outgoing context metadata if propagation direction is outgoing,
+// otherwise it returns keys from incoming context metadata.
 func (c *CustomCarrier) Keys() []string {
-	var keys []string
-	md, ok := metadata.FromOutgoingContext(c.ctx)
-	if ok {
-		keys = make([]string, 0, len(md))
-		for key := range md {
-			keys = append(keys, key)
-		}
+	var md metadata.MD
+	var ok bool
+
+	switch c.direction {
+	case Outgoing:
+		md, ok = metadata.FromOutgoingContext(c.ctx)
+	case Incoming:
+		md, ok = metadata.FromIncomingContext(c.ctx)
+	default:
```

Comment:
You don't need this case since you check `ok` below.

Reply:
not anymore because with two new incoming and outgoing constructors. However, it was applicable before when we were accepting direction from user